### PR TITLE
fix(cdk-experimental/menu): do not allow two separate triggers to open the same menu

### PR DESCRIPTION
**Pull request by andy9775**
Fixes the error where two different menu triggers open the same menu which leads to open menus not
closing out on hover and other inconsistent behaviour. Opt to throw and error and not allow two
triggers to share a menu.

### CODE REVIEW

File: src/cdk-experimental/menu/menu-item-trigger.ts [hunk 1]
```diff
@@ -140,7 +148,7 @@ export class CdkMenuItemTrigger implements OnDestroy {
    */
   _toggleOnMouseEnter() {
     const menuStack = this._getMenuStack();
-    if (!menuStack.isEmpty() && !this.isMenuOpen()) {
+    if (!menuStack?.isEmpty() && !this.isMenuOpen()) {
```
Comment (teflonwaffles) [inline src/cdk-experimental/menu/menu-item-trigger.ts:151]: The first part of this expression is a little hard to parse. Is there a way to make it clearer?

File: src/cdk-experimental/menu/menu-item-trigger.ts [hunk 2]
```diff
@@ -206,10 +214,10 @@ export class CdkMenuItemTrigger implements OnDestroy {
     // that means that the parent menu is a menu bar since we don't put the menu bar on the
     // stack
     const isParentMenuBar =
-      !menuStack.closeSubMenuOf(this._parentMenu) && menuStack.peek() !== this._parentMenu;
+      !menuStack?.closeSubMenuOf(this._parentMenu) && menuStack?.peek() !== this._parentMenu;
```
Comment (teflonwaffles) [inline src/cdk-experimental/menu/menu-item-trigger.ts:217]: Could you verify that your calls to methods on menuStack? will work in an expected way if menuStack is not defined?

File: src/cdk-experimental/menu/context-menu.ts [hunk 3]
```diff
@@ -108,6 +110,10 @@ export class CdkContextMenuTrigger implements OnDestroy {
     return this._menuPanel;
   }
   set menuPanel(panel: CdkMenuPanel) {
+    // If the provided panel already has a stack, that means it already has a trigger configured
+    if (panel._menuStack && isDevMode()) {
```
Comment (jelbourn) [inline src/cdk-experimental/menu/context-menu.ts:114]: I don't think it makes a difference, but conventionally the `isDevMode` check goes first

Comment (crisbeto) [inline src/cdk-experimental/menu/context-menu.ts:114]: Note that the `isDevMode` calls will become lint errors with https://github.com/angular/components/pull/20146. I suspect that your changes will land first though.

File: src/cdk-experimental/menu/menu-item-trigger.spec.ts [hunk 4]
```diff
@@ -293,6 +301,35 @@ describe('MenuItemTrigger', () => {
...
+    it('should throw an error if two triggers in different menubars open the same menu', () => {
+      expect(createComponent(TriggersWithSameMenuDifferentMenuBars)).toThrowError();
+    });
+
+    it('should throw an error if two triggers in the same menubar open the same menu', () => {
+      expect(createComponent(TriggersWithSameMenuSameMenuBar)).toThrowError();
+    });
+
+    it('should throw an error if a trigger in a submenu references its parent menu', () => {
+      expect(createComponent(TriggerOpensItsMenu)).toThrowError();
```
Comment (crisbeto) [inline src/cdk-experimental/menu/menu-item-trigger.spec.ts:330]: Should we be testing for the specific error that we're expecting here? E.g. `.toThrowError(/CdkMenuPanel is already referenced by different CdkMenuTrigger/)`.

File: src/cdk-experimental/menu/menu-item.ts [hunk 5]
```diff
@@ -125,7 +125,7 @@ export class CdkMenuItem implements FocusableOption, FocusableElement, OnDestroy
     }
 
     // don't set the tabindex if there are no open sibling or parent menus
-    if (!event || (event && !this._getMenuStack().isEmpty())) {
+    if (!event || (event && !this._getMenuStack()?.isEmpty())) {
```
Comment (crisbeto) [inline src/cdk-experimental/menu/menu-item.ts:128]: I think the `event &&` here is redundant since it's assumed by the `!event` being true before.

File: src/cdk-experimental/menu/context-menu.ts (hunk 3 above)
Comment (andy9775) [reply src/cdk-experimental/menu/context-menu.ts:114]: ya, I saw that and wasn't sure how to address it. I think I'll issue a follow up PR after it lands - I'll add a TODO for now

File: src/cdk-experimental/menu/menu-item-trigger.ts (hunk 2 above)
Comment (andy9775) [reply src/cdk-experimental/menu/menu-item-trigger.ts:217]: In the instance that there is no menustack there's nothing to do and the `?` prevents the methods from being called.

**Comment by andy9775** [comment]
@jelbourn @teflonwaffles @crisbeto feedback should be addressed

**Comment by JonWallsten** [comment]
@andy9775: Is this going to affect the Material Menu and Material Menu Trigger as well? We are actually using two different triggers for the same menu with with a purpose. We have the classic tre-dot-menu but also show the menu when using right-click, only then we want the menu to appear where user is clicking. Since the easiest way to get the menu position to be where you click at is to place a trigger there, we have another trigger for that. With only one trigger we would have to move the dots to where the mouse is. Which obviously doesn't look that good. 
Worst case, could this feature be optional?

**Comment by andy9775** [comment]
@JonWallsten this pr only impacts the cdk-experimental/menu package and not material.

**Comment by jelbourn** [comment]
@JonWallsten in terms of the future- I want to attempt to land a change in Angular that would make this restriction unnecessary before changing the interns of `MatMenu`
